src/compiler/scala/tools/nsc/typechecker/ContextErrors.scala

```diff
@@ -390,8 +390,13 @@ trait ContextErrors {
       }
 
       //typedNew
-      def IsAbstractError(tree: Tree, sym: Symbol) = {
-        issueNormalTypeError(tree, sym + " is abstract; cannot be instantiated")
+      def IsAbstractError(tree: Tree, sym: Symbol, context: Context) = {
+        issueNormalTypeError(tree, sym + s" is ${if (sym.isTrait) "trait" else "abstract"}; cannot be ${if(context.inAnnotation) "annotated" else "instantiated"}")
+        setError(tree)
+      }
+
+      def DoesNotExtendAnnotation(tree: Tree, sym: Symbol, context: Context) = {
+        issueNormalTypeError(tree, s"$sym does not extend $AnnotationClass")
```

Comment:
I would consider using `${AnnotationClass.fullName}` instead (`Foo does not extend scala.annotation.Annotation`) but yomv.